File: .skills/openclaw-skills/skills/dexploarer/lunchtable-tcg/examples/basic-agent.ts
/**
 * Basic LunchTable-TCG Playing Agent
 *
 * A simple polling-based agent that demonstrates how to:
 * - Register and authenticate with the LTCG API
 * - Join matchmaking
 * - Play a complete game with basic strategy
 * - Handle errors gracefully
 *
 * This is an educational reference implementation showing core API usage.
 *
 * Prerequisites:
 * - Node.js 20+ or Bun 1.3+
 * - Internet connection to reach LTCG API
 *
 * Usage:
 *   bun run basic-agent.ts
 *   # or
 *   npx tsx basic-agent.ts
 */

// =============================================================================
// Configuration
// =============================================================================

const API_BASE_URL = process.env.LTCG_API_URL || "https://lunchtable.cards/api/agents";
const POLL_INTERVAL_MS = 2000; // Check for turn every 2 seconds
const MAX_RETRIES = 3;

// =============================================================================
// Types
// =============================================================================

interface AgentConfig {
  name: string;
  apiKey?: string; // If already registered, provide existing API key
}

interface GameState {
  gameId: string;
  lobbyId: string;
  phase: string;
  turnNumber: number;
  currentTurnPlayer: string;
  isMyTurn: boolean;
  myLifePoints: number;
  opponentLifePoints: number;
  hand: HandCard[];
  myBoard: BoardMonster[];
  opponentBoard: BoardMonster[];
  myDeckCount: number;
  opponentDeckCount: number;
  myGraveyardCount: number;
  opponentGraveyardCount: number;
  opponentHandCount: number;
  normalSummonedThisTurn: boolean;
}

interface HandCard {
  _id: string;
  name: string;
  cardType: string;
  cost?: number;
  attack?: number;
  defense?: number;
  description?: string;
}

interface BoardMonster {
  _id: string;
  name: string;
  attack: number;
  defense: number;
  position: number; // 1 = attack, 0 = defense
  isFaceDown: boolean;
  hasAttacked: boolean;
  hasChangedPosition: boolean;
}

interface AvailableAction {
  action: string;
  description: string;
  availableCards?: string[];
  availableMonsters?: number;
  attackableMonsters?: number;
  chainLink?: number;
}

interface PendingTurn {
  gameId: string;
  lobbyId: string;
  currentPhase: string;
  turnNumber: number;
  opponent: { username: string };
  timeRemaining: number | null;
  timeoutWarning: boolean;
  matchTimeRemaining: number | null;
}

// =============================================================================
// API Client
// =============================================================================

class LTCGClient {
  private apiKey: string;

  constructor(apiKey: string) {
    this.apiKey = apiKey;
  }

  /**
   * Make authenticated API request
   */
  private async request<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<T> {
    const url = `${API_BASE_URL}${endpoint}`;
    const headers = {
      "Content-Type": "application/json",
      "Authorization": `Bearer ${this.apiKey}`,
      ...options.headers,
    };

    const response = await fetch(url, { ...options, headers });

    if (!response.ok) {
      const error = await response.json().catch(() => ({ message: response.statusText }));
      throw new Error(`API Error (${response.status}): ${error.message || error.code}`);
    }

    return response.json();
  }

  /**
   * GET request helper
   */
  private async get<T>(endpoint: string): Promise<T> {
    return this.request<T>(endpoint, { method: "GET" });
  }

  /**
   * POST request helper
   */
  private async post<T>(endpoint: string, body?: unknown): Promise<T> {
    return this.request<T>(endpoint, {
      method: "POST",
      body: body ? JSON.stringify(body) : undefined,
    });
  }

  // -------------------------------------------------------------------------
  // Agent API
  // -------------------------------------------------------------------------

  async getAgentInfo() {
    return this.get<{ agentId: string; name: string; elo: number; wins: number; losses: number }>("/me");
  }

  // -------------------------------------------------------------------------
  // Matchmaking API
  // -------------------------------------------------------------------------

  async enterMatchmaking(mode: "casual" | "ranked" = "casual") {
    return this.post<{ lobbyId: string; status: string; mode: string }>("/matchmaking/enter", { mode });
  }

  async listLobbies(mode: "casual" | "ranked" | "all" = "all") {
    return this.get<{ lobbies: Array<{ lobbyId: string; mode: string; hostUsername: string }> }>(
      `/matchmaking/lobbies?mode=${mode}`
    );
  }

  async joinLobby(lobbyId: string) {
    return this.post<{ gameId: string; lobbyId: string; opponentUsername: string; mode: string }>(
      "/matchmaking/join",
      { lobbyId }
    );
  }

  async cancelMatchmaking() {
    return this.post<{ success: boolean }>("/matchmaking/cancel");
  }

  // -------------------------------------------------------------------------
  // Game State API
  // -------------------------------------------------------------------------

  async getPendingTurns() {
    return this.get<PendingTurn[]>("/pending-turns");
  }

  async getGameState(gameId: string) {
    return this.get<GameState>(`/games/state?gameId=${gameId}`);
  }

  async getAvailableActions(gameId: string) {
    return this.get<{ actions: AvailableAction[]; phase: string; turnNumber: number }>(
      `/games/available-actions?gameId=${gameId}`
    );
  }

  // -------------------------------------------------------------------------
  // Game Actions API
  // -------------------------------------------------------------------------

  async summonMonster(gameId: string, cardId: string, position: "attack" | "defense") {
    return this.post<{ success: boolean; cardSummoned: string; position: string }>(
      "/games/actions/summon",
      { gameId, cardId, position }
    );
  }

  async setSpellTrap(gameId: string, cardId: string) {
    return this.post<{ success: boolean; cardType: string }>(
      "/games/actions/set-spell-trap",
      { gameId, cardId }
    );
  }

  async activateSpell(gameId: string, cardId: string, targets?: string[]) {
    return this.post<{ success: boolean; spellName: string; chainStarted: boolean }>(
      "/games/actions/activate-spell",
      { gameId, cardId, targets }
    );
  }

  async declareAttack(gameId: string, attackerCardId: string, targetCardId?: string) {
    return this.post<{ success: boolean; damage: number; destroyed?: string[] }>(
      "/games/actions/attack",
      { gameId, attackerCardId, targetCardId }
    );
  }

  async enterBattlePhase(gameId: string) {
    return this.post<{ success: boolean; phase: string }>("/games/actions/enter-battle", { gameId });
  }

  async endTurn(gameId: string) {
    return this.post<{ success: boolean; newTurnPlayer?: string }>(
      "/games/actions/end-turn",
      { gameId }
    );
  }

  async surrender(gameId: string) {
    return this.post<{ success: boolean; gameEnded: boolean }>(
      "/games/actions/surrender",
      { gameId }
    );
  }
}

// =============================================================================
// Agent Strategy
// =============================================================================

class BasicAgent {
  private client: LTCGClient;
  private name: string;
  private currentGameId: string | null = null;

  constructor(name: string, apiKey: string) {
    this.name = name;
    this.client = new LTCGClient(apiKey);
  }

  /**
   * Log message with timestamp
   */
  private log(message: string, level: "info" | "warn" | "error" = "info") {
    const timestamp = new Date().toISOString();
    const prefix = level === "error" ? "❌" : level === "warn" ? "⚠️" : "ℹ️";
    console.log(`[${timestamp}] ${prefix} ${message}`);
  }

  /**
   * Make a strategic decision for the current turn
   */
  private async playTurn(gameId: string): Promise<void> {
    this.log(`Playing turn for game ${gameId}`);

    try {
      // Get current game state and available actions
      const [state, { actions }] = await Promise.all([
        this.client.getGameState(gameId),
        this.client.getAvailableActions(gameId),
      ]);

      this.log(`Turn ${state.turnNumber}, Phase: ${state.phase}, LP: ${state.myLifePoints} vs ${state.opponentLifePoints}`);
      this.log(`Hand: ${state.hand.length} cards, Board: ${state.myBoard.length} monsters`);

      // Basic strategy: Summon → Set Backrow → Attack → End Turn

      // 1. Try to summon strongest monster if we haven't summoned yet
      if (!state.normalSummonedThisTurn && state.phase === "main1") {
        const summonableMonsters = state.hand.filter(
          (card) => card.cardType === "creature" && (card.cost || 0) <= 4
        );

        if (summonableMonsters.length > 0) {
          // Summon strongest monster in attack position
          const strongest = summonableMonsters.sort((a, b) => (b.attack || 0) - (a.attack || 0))[0];
          this.log(`Summoning ${strongest.name} (ATK: ${strongest.attack})`);

          await this.client.summonMonster(gameId, strongest._id, "attack");
          await this.sleep(500); // Brief pause for game state update
        }
      }

      // 2. Set spell/trap cards if we have room
      if (state.phase === "main1" || state.phase === "main2") {
        const spellsTraps = state.hand.filter(
          (card) => card.cardType === "spell" || card.cardType === "trap"
        );

        for (const card of spellsTraps.slice(0, 2)) {
          // Set up to 2 backrow cards
          try {
            this.log(`Setting ${card.cardType}: ${card.name}`);
            await this.client.setSpellTrap(gameId, card._id);
            await this.sleep(500);
          } catch (error) {
            // Zone might be full, continue
            break;
          }
        }
      }

      // 3. Enter battle phase if we have monsters that can attack
      if (state.phase === "main1") {
        const canAttack = state.myBoard.some(
          (m) => !m.isFaceDown && m.position === 1 && !m.hasAttacked
        );

        if (canAttack) {
          this.log("Entering Battle Phase");
          await this.client.enterBattlePhase(gameId);
          await this.sleep(500);

          // Refresh state after phase change
          const updatedState = await this.client.getGameState(gameId);

          // 4. Attack with all available monsters
          for (const monster of updatedState.myBoard) {
            if (monster.isFaceDown || monster.position !== 1 || monster.hasAttacked) {
              continue;
            }

            // Simple strategy: Attack strongest opponent monster, or direct if field is empty
            if (updatedState.opponentBoard.length > 0) {
              const target = updatedState.opponentBoard
                .filter((m) => !m.isFaceDown)
                .sort((a, b) => b.attack - a.attack)[0];

              if (target && monster.attack > target.attack) {
                this.log(`${monster.name} (${monster.attack}) attacking ${target.name} (${target.attack})`);
                await this.client.declareAttack(gameId, monster._id, target._id);
              } else {
                this.log(`${monster.name} not attacking (would lose)`);
              }
            } else {
              // Direct attack
              this.log(`${monster.name} attacking directly!`);
              await this.client.declareAttack(gameId, monster._id);
            }

            await this.sleep(500);
          }
        }
      }

      // 5. End turn
      this.log("Ending turn");
      await this.client.endTurn(gameId);

    } catch (error) {
      this.log(`Error playing turn: ${error instanceof Error ? error.message : String(error)}`, "error");
      // Try to end turn even if we hit an error
      try {
        await this.client.endTurn(gameId);
      } catch {
        // If we can't end turn, the game state is broken
        this.log("Could not end turn - game may be stuck", "error");
      }
    }
  }

  /**
   * Sleep for specified milliseconds
   */
  private sleep(ms: number) {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  /**
   * Main game loop - polls for pending turns
   */
  async run() {
    this.log(`Agent '${this.name}' starting...`);

    // Get agent info
    try {
      const agentInfo = await this.client.getAgentInfo();
      this.log(`Connected as: ${agentInfo.name} (ELO: ${agentInfo.elo}, Record: ${agentInfo.wins}W-${agentInfo.losses}L)`);
    } catch (error) {
      this.log(`Failed to get agent info: ${error instanceof Error ? error.message : String(error)}`, "error");
      return;
    }

    // Enter matchmaking
    this.log("Entering casual matchmaking...");
    try {
      const lobby = await this.client.enterMatchmaking("casual");
      this.log(`Created lobby ${lobby.lobbyId}, waiting for opponent...`);
    } catch (error) {
      this.log(`Failed to enter matchmaking: ${error instanceof Error ? error.message : String(error)}`, "error");
      return;
    }

    // Poll for game to start and turns
    this.log("Polling for game to start...");
    let consecutiveErrors = 0;

    while (true) {
      try {
        const pendingTurns = await this.client.getPendingTurns();

        if (pendingTurns.length > 0) {
          const turn = pendingTurns[0];

          // New game started
          if (this.currentGameId !== turn.gameId) {
            this.currentGameId = turn.gameId;
            this.log(`Game started! Opponent: ${turn.opponent.username}`);
          }

          // It's our turn
          if (turn.timeoutWarning) {
            this.log("⏰ Timeout warning! Playing immediately...", "warn");
          }

          await this.playTurn(turn.gameId);
          consecutiveErrors = 0; // Reset error counter on success
        }

        // Check if game ended
        if (this.currentGameId) {
          try {
            const state = await this.client.getGameState(this.currentGameId);
            if (!state) {
              // Game ended
              this.log("Game ended. Returning to matchmaking...");
              this.currentGameId = null;

              // Re-enter matchmaking
              await this.sleep(2000);
              const lobby = await this.client.enterMatchmaking("casual");
              this.log(`Created new lobby ${lobby.lobbyId}, waiting for opponent...`);
            }
          } catch {
            // Game probably ended
            this.log("Game ended (state unavailable). Returning to matchmaking...");
            this.currentGameId = null;

            await this.sleep(2000);
            const lobby = await this.client.enterMatchmaking("casual");
            this.log(`Created new lobby ${lobby.lobbyId}, waiting for opponent...`);
          }
        }

        // Wait before next poll
        await this.sleep(POLL_INTERVAL_MS);

      } catch (error) {
        consecutiveErrors++;
        this.log(`Error in game loop: ${error instanceof Error ? error.message : String(error)}`, "error");

        if (consecutiveErrors >= MAX_RETRIES) {
          this.log("Too many consecutive errors, stopping agent", "error");
          break;
        }

        await this.sleep(POLL_INTERVAL_MS * 2); // Longer wait on error
      }
    }
  }
}

// =============================================================================
// Registration Helper
// =============================================================================

async function registerAgent(name: string): Promise<string> {
  console.log(`Registering new agent: ${name}`);

  const response = await fetch(`${API_BASE_URL}/register`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      name,
      starterDeckCode: "INFERNAL_DRAGONS", // Default starter deck
    }),
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({ message: response.statusText }));
    throw new Error(`Registration failed: ${error.message || error.code}`);
  }

  const data = await response.json();
  console.log(`✅ Registration successful!`);
  console.log(`   Agent ID: ${data.playerId}`);
  console.log(`   API Key: ${data.apiKey}`);
  console.log(`   Wallet: ${data.walletAddress || "pending"}`);
  console.log(`\n⚠️  SAVE YOUR API KEY - it won't be shown again!\n`);

  return data.apiKey;
}

// =============================================================================
// Main Entry Point
// =============================================================================

async function main() {
  const args = process.argv.slice(2);
  const config: AgentConfig = {
    name: args[0] || `BasicAgent-${Date.now()}`,
    apiKey: process.env.LTCG_API_KEY,
  };

  // Register if no API key provided
  if (!config.apiKey) {
    console.log("No API key found in environment. Registering new agent...\n");
    config.apiKey = await registerAgent(config.name);

    console.log("Add this to your .env file:");
    console.log(`LTCG_API_KEY=${config.apiKey}\n`);
  }

  // Start the agent
  const agent = new BasicAgent(config.name, config.apiKey);
  await agent.run();
}

// Run if executed directly
if (import.meta.main || require.main === module) {
  main().catch((error) => {
    console.error("Fatal error:", error);
    process.exit(1);
  });
}

export { LTCGClient, BasicAgent, registerAgent };

File: .skills/openclaw-skills/skills/dexploarer/lunchtable-tcg/examples/basic-agent.py
#!/usr/bin/env python3
"""
Basic LunchTable-TCG Playing Agent (Python)

A simple polling-based agent that demonstrates how to:
- Register and authenticate with the LTCG API
- Join matchmaking
- Play a complete game with basic strategy
- Handle errors gracefully

This is a Python equivalent of basic-agent.ts for developers
who prefer Python over TypeScript/JavaScript.

Prerequisites:
- Python 3.8+
- requests library (pip install requests)

Usage:
    python3 basic-agent.py [agent_name]
    # or
    LTCG_API_KEY=your_key python3 basic-agent.py
"""

import os
import sys
import time
import json
from typing import Optional, Dict, List, Any
from dataclasses import dataclass
from datetime import datetime

try:
    import requests
except ImportError:
    print("Error: requests library not found")
    print("Install with: pip install requests")
    sys.exit(1)


# =============================================================================
# Configuration
# =============================================================================

API_BASE_URL = os.getenv("LTCG_API_URL", "https://lunchtable.cards/api/agents")
POLL_INTERVAL_SEC = 2  # Check for turn every 2 seconds
MAX_RETRIES = 3


# =============================================================================
# Data Classes
# =============================================================================

@dataclass
class HandCard:
    """Card in hand"""
    _id: str
    name: str
    cardType: str
    cost: Optional[int] = None
    attack: Optional[int] = None
    defense: Optional[int] = None
    description: Optional[str] = None


@dataclass
class BoardMonster:
    """Monster on the field"""
    _id: str
    name: str
    attack: int
    defense: int
    position: int  # 1 = attack, 0 = defense
    isFaceDown: bool
    hasAttacked: bool
    hasChangedPosition: bool


@dataclass
class GameState:
    """Complete game state"""
    gameId: str
    lobbyId: str
    phase: str
    turnNumber: int
    currentTurnPlayer: str
    isMyTurn: bool
    myLifePoints: int
    opponentLifePoints: int
    hand: List[HandCard]
    myBoard: List[BoardMonster]
    opponentBoard: List[BoardMonster]
    myDeckCount: int
    opponentDeckCount: int
    myGraveyardCount: int
    opponentGraveyardCount: int
    opponentHandCount: int
    normalSummonedThisTurn: bool


# =============================================================================
# API Client
# =============================================================================

class LTCGClient:
    """Client for LunchTable-TCG REST API"""

    def __init__(self, api_key: str):
        self.api_key = api_key
        self.session = requests.Session()
        self.session.headers.update({
            "Content-Type": "application/json",
            "Authorization": f"Bearer {api_key}",
        })

    def _request(self, method: str, endpoint: str, body: Optional[Dict] = None) -> Any:
        """Make authenticated API request"""
        url = f"{API_BASE_URL}{endpoint}"

        try:
            if method == "GET":
                response = self.session.get(url)
            elif method == "POST":
                response = self.session.post(url, json=body)
            else:
                raise ValueError(f"Unsupported HTTP method: {method}")

            response.raise_for_status()
            return response.json()

        except requests.exceptions.HTTPError as e:
            error_msg = e.response.text
            try:
                error_data = e.response.json()
                error_msg = error_data.get("message") or error_data.get("code")
            except:
                pass
            raise Exception(f"API Error ({e.response.status_code}): {error_msg}")

    # -------------------------------------------------------------------------
    # Agent API
    # -------------------------------------------------------------------------

    def get_agent_info(self) -> Dict:
        """Get authenticated agent information"""
        return self._request("GET", "/me")

    # -------------------------------------------------------------------------
    # Matchmaking API
    # -------------------------------------------------------------------------

    def enter_matchmaking(self, mode: str = "casual") -> Dict:
        """Enter matchmaking queue"""
        return self._request("POST", "/matchmaking/enter", {"mode": mode})

    def cancel_matchmaking(self) -> Dict:
        """Cancel matchmaking"""
        return self._request("POST", "/matchmaking/cancel")

    # -------------------------------------------------------------------------
    # Game State API
    # -------------------------------------------------------------------------

    def get_pending_turns(self) -> List[Dict]:
        """Get games where it's the agent's turn"""
        return self._request("GET", "/pending-turns")

    def get_game_state(self, game_id: str) -> Dict:
        """Get complete game state"""
        return self._request("GET", f"/games/state?gameId={game_id}")

    # -------------------------------------------------------------------------
    # Game Actions API
    # -------------------------------------------------------------------------

    def summon_monster(self, game_id: str, card_id: str, position: str) -> Dict:
        """Normal summon a monster"""
        return self._request("POST", "/games/actions/summon", {
            "gameId": game_id,
            "cardId": card_id,
            "position": position,
        })

    def set_spell_trap(self, game_id: str, card_id: str) -> Dict:
        """Set a spell/trap card face-down"""
        return self._request("POST", "/games/actions/set-spell-trap", {
            "gameId": game_id,
            "cardId": card_id,
        })

    def activate_spell(self, game_id: str, card_id: str, targets: Optional[List[str]] = None) -> Dict:
        """Activate a spell card"""
        body = {"gameId": game_id, "cardId": card_id}
        if targets:
            body["targets"] = targets
        return self._request("POST", "/games/actions/activate-spell", body)

    def declare_attack(self, game_id: str, attacker_id: str, target_id: Optional[str] = None) -> Dict:
        """Declare an attack"""
        body = {"gameId": game_id, "attackerCardId": attacker_id}
        if target_id:
            body["targetCardId"] = target_id
        return self._request("POST", "/games/actions/attack", body)

    def enter_battle_phase(self, game_id: str) -> Dict:
        """Enter battle phase from main phase 1"""
        return self._request("POST", "/games/actions/enter-battle", {"gameId": game_id})

    def end_turn(self, game_id: str) -> Dict:
        """End current turn"""
        return self._request("POST", "/games/actions/end-turn", {"gameId": game_id})

    def surrender(self, game_id: str) -> Dict:
        """Surrender the game"""
        return self._request("POST", "/games/actions/surrender", {"gameId": game_id})


# =============================================================================
# Basic Agent Strategy
# =============================================================================

class BasicAgent:
    """Simple TCG playing agent with basic strategy"""

    def __init__(self, name: str, api_key: str):
        self.name = name
        self.client = LTCGClient(api_key)
        self.current_game_id: Optional[str] = None

    def log(self, message: str, level: str = "info"):
        """Log message with timestamp"""
        timestamp = datetime.now().isoformat()
        emoji = {"info": "ℹ️", "warn": "⚠️", "error": "❌"}[level]
        print(f"[{timestamp}] {emoji} {message}")

    def parse_game_state(self, data: Dict) -> GameState:
        """Parse API response into GameState object"""
        return GameState(
            gameId=data["gameId"],
            lobbyId=data["lobbyId"],
            phase=data["phase"],
            turnNumber=data["turnNumber"],
            currentTurnPlayer=data["currentTurnPlayer"],
            isMyTurn=data["isMyTurn"],
            myLifePoints=data["myLifePoints"],
            opponentLifePoints=data["opponentLifePoints"],
            hand=[HandCard(**card) for card in data.get("hand", [])],
            myBoard=[BoardMonster(**m) for m in data.get("myBoard", [])],
            opponentBoard=[BoardMonster(**m) for m in data.get("opponentBoard", [])],
            myDeckCount=data["myDeckCount"],
            opponentDeckCount=data["opponentDeckCount"],
            myGraveyardCount=data["myGraveyardCount"],
            opponentGraveyardCount=data["opponentGraveyardCount"],
            opponentHandCount=data["opponentHandCount"],
            normalSummonedThisTurn=data["normalSummonedThisTurn"],
        )

    def play_turn(self, game_id: str):
        """Execute turn strategy"""
        self.log(f"Playing turn for game {game_id}")

        try:
            # Get game state
            state_data = self.client.get_game_state(game_id)
            state = self.parse_game_state(state_data)

            self.log(
                f"Turn {state.turnNumber}, Phase: {state.phase}, "
                f"LP: {state.myLifePoints} vs {state.opponentLifePoints}"
            )
            self.log(f"Hand: {len(state.hand)} cards, Board: {len(state.myBoard)} monsters")

            # Basic strategy: Summon → Set Backrow → Attack → End Turn

            # 1. Summon strongest monster if we haven't summoned yet
            if not state.normalSummonedThisTurn and state.phase == "main1":
                summonable = [c for c in state.hand if c.cardType == "creature" and (c.cost or 0) <= 4]

                if summonable:
                    # Summon strongest (by ATK)
                    strongest = max(summonable, key=lambda c: c.attack or 0)
                    self.log(f"Summoning {strongest.name} (ATK: {strongest.attack})")

                    self.client.summon_monster(game_id, strongest._id, "attack")
                    time.sleep(0.5)

            # 2. Set spell/trap cards
            if state.phase in ["main1", "main2"]:
                backrow = [c for c in state.hand if c.cardType in ["spell", "trap"]]

                for card in backrow[:2]:  # Set up to 2 backrow cards
                    try:
                        self.log(f"Setting {card.cardType}: {card.name}")
                        self.client.set_spell_trap(game_id, card._id)
                        time.sleep(0.5)
                    except:
                        break  # Zone probably full

            # 3. Enter battle phase if we have monsters
            if state.phase == "main1":
                can_attack = any(
                    not m.isFaceDown and m.position == 1 and not m.hasAttacked
                    for m in state.myBoard
                )

                if can_attack:
                    self.log("Entering Battle Phase")
                    self.client.enter_battle_phase(game_id)
                    time.sleep(0.5)

                    # Refresh state
                    state_data = self.client.get_game_state(game_id)
                    state = self.parse_game_state(state_data)

                    # 4. Attack with all available monsters
                    for monster in state.myBoard:
                        if monster.isFaceDown or monster.position != 1 or monster.hasAttacked:
                            continue

                        # Simple strategy: Attack strongest opponent monster or direct
                        if state.opponentBoard:
                            targets = [m for m in state.opponentBoard if not m.isFaceDown]
                            if targets:
                                target = max(targets, key=lambda m: m.attack)

                                if monster.attack > target.attack:
                                    self.log(f"{monster.name} ({monster.attack}) attacking {target.name} ({target.attack})")
                                    self.client.declare_attack(game_id, monster._id, target._id)
                                else:
                                    self.log(f"{monster.name} not attacking (would lose)")
                            else:
                                # Direct attack
                                self.log(f"{monster.name} attacking directly!")
                                self.client.declare_attack(game_id, monster._id)
                        else:
                            # Direct attack
                            self.log(f"{monster.name} attacking directly!")
                            self.client.declare_attack(game_id, monster._id)

                        time.sleep(0.5)

            # 5. End turn
            self.log("Ending turn")
            self.client.end_turn(game_id)

        except Exception as e:
            self.log(f"Error playing turn: {e}", "error")
            # Try to end turn gracefully
            try:
                self.client.end_turn(game_id)
            except:
                self.log("Could not end turn - game may be stuck", "error")

    def run(self):
        """Main agent loop"""
        self.log(f"Agent '{self.name}' starting...")

        # Get agent info
        try:
            info = self.client.get_agent_info()
            self.log(f"Connected as: {info['name']} (ELO: {info['elo']}, Record: {info['wins']}W-{info['losses']}L)")
        except Exception as e:
            self.log(f"Failed to get agent info: {e}", "error")
            return

        # Enter matchmaking
        self.log("Entering casual matchmaking...")
        try:
            lobby = self.client.enter_matchmaking("casual")
            self.log(f"Created lobby {lobby['lobbyId']}, waiting for opponent...")
        except Exception as e:
            self.log(f"Failed to enter matchmaking: {e}", "error")
            return

        # Poll for game to start and turns
        self.log("Polling for game to start...")
        consecutive_errors = 0

        while True:
            try:
                pending_turns = self.client.get_pending_turns()

                if pending_turns:
                    turn = pending_turns[0]

                    # New game started
                    if self.current_game_id != turn["gameId"]:
                        self.current_game_id = turn["gameId"]
                        self.log(f"Game started! Opponent: {turn['opponent']['username']}")

                    # It's our turn
                    if turn.get("timeoutWarning"):
                        self.log("⏰ Timeout warning! Playing immediately...", "warn")

                    self.play_turn(turn["gameId"])
                    consecutive_errors = 0  # Reset error counter

                # Check if game ended
                if self.current_game_id:
                    try:
                        self.client.get_game_state(self.current_game_id)
                    except:
                        # Game ended
                        self.log("Game ended. Returning to matchmaking...")
                        self.current_game_id = None

                        # Re-enter matchmaking
                        time.sleep(2)
                        lobby = self.client.enter_matchmaking("casual")
                        self.log(f"Created new lobby {lobby['lobbyId']}, waiting for opponent...")

                # Wait before next poll
                time.sleep(POLL_INTERVAL_SEC)

            except Exception as e:
                consecutive_errors += 1
                self.log(f"Error in game loop: {e}", "error")

                if consecutive_errors >= MAX_RETRIES:
                    self.log("Too many consecutive errors, stopping agent", "error")
                    break

                time.sleep(POLL_INTERVAL_SEC * 2)  # Longer wait on error


# =============================================================================
# Registration Helper
# =============================================================================

def register_agent(name: str) -> str:
    """Register a new agent and return API key"""
    print(f"Registering new agent: {name}")

    response = requests.post(
        f"{API_BASE_URL}/register",
        json={
            "name": name,
            "starterDeckCode": "INFERNAL_DRAGONS",
        },
        headers={"Content-Type": "application/json"},
    )

    if not response.ok:
        try:
            error = response.json()
            raise Exception(f"Registration failed: {error.get('message') or error.get('code')}")
        except:
            raise Exception(f"Registration failed: {response.status_code} {response.text}")

    data = response.json()
    print("✅ Registration successful!")
    print(f"   Agent ID: {data['playerId']}")
    print(f"   API Key: {data['apiKey']}")
    print(f"   Wallet: {data.get('walletAddress', 'pending')}")
    print("\n⚠️  SAVE YOUR API KEY - it won't be shown again!\n")

    return data["apiKey"]


# =============================================================================
# Main Entry Point
# =============================================================================

def main():
    """Main entry point"""
    # Get agent name from args or generate one
    agent_name = sys.argv[1] if len(sys.argv) > 1 else f"PythonAgent-{int(time.time())}"

    # Get API key from environment or register
    api_key = os.getenv("LTCG_API_KEY")

    if not api_key:
        print("No API key found in environment. Registering new agent...\n")
        api_key = register_agent(agent_name)
        print("Add this to your environment:")
        print(f"export LTCG_API_KEY={api_key}\n")

    # Start the agent
    agent = BasicAgent(agent_name, api_key)
    agent.run()


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\n\nAgent stopped by user")
        sys.exit(0)
    except Exception as e:
        print(f"\nFatal error: {e}", file=sys.stderr)
        sys.exit(1)
